File: testsuites/HLT/Benchmark/micro-benchmark/swift/loop/BenchmarkWhile.swift
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

import Foundation

let reps = 100000000
var data = 0

extension TimeInterval {
    func formatted(decimalPoint: Int) -> String {
        if (decimalPoint == 2) {
            return String(format: "%.2f", self)
        }
        return String(format: "%.3f", self)
    }
}

@inline(never)
func BenchmarkWhile(x: Int) -> TimeInterval {
    var i = 0
    var totalTime: Double = 0.0

    for _ in 0..<reps {
        let startTime = DispatchTime.now()

        while i < x {
            data += 1
            i += 1
         }

        let endTime = DispatchTime.now()
        totalTime += Double(endTime.uptimeNanoseconds - startTime.uptimeNanoseconds)
        data = 0
    }

    let perTime = totalTime / Double(reps)

    return perTime
}

print("BenchmarkWhile_N_1K: \(BenchmarkWhile(x: 1000).formatted(decimalPoint:2)) ns/op")
print("BenchmarkWhile_N_1B: \(BenchmarkWhile(x: 1000_000).formatted(decimalPoint:2)) ns/op")
print("BenchmarkWhile_N_1M: \(BenchmarkWhile(x: 1000_000_000).formatted(decimalPoint:2)) ns/op")

File: testsuites/HLT/Benchmark/micro-benchmark/java/expression/BenchmarkWhile.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

package expression;

import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.concurrent.TimeUnit;

@BenchmarkMode(Mode.AverageTime)
@Warmup(iterations = 1, time = 1)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Measurement(iterations = 1)
@State(Scope.Benchmark)
@Fork(1)

public class BenchmarkWhile {
    static int sum = 0;

    @Benchmark
    public void BenchmarkWhileD1(Blackhole blackhole) {
        int n = 0;
        sum = 0;
        while (n < 100000) {
            n++;
            sum += n / 2;
        }
        blackhole.consume(sum);
        blackhole.consume(n);
    }

    @Benchmark
    public void BenchmarkWhileD2(Blackhole blackhole) {
        int n, m;
        n = m = 0;
        sum = 0;
        while (n < 1000) {
            n++;
            m = 0;
            while (m < 100) {
                m++;
                sum += (n + m) / 2;
            }
        }
        blackhole.consume(sum);
        blackhole.consume(n);
        blackhole.consume(m);
    }

    @Benchmark
    public void BenchmarkWhileD3(Blackhole blackhole) {
        int n, m, j;
        n = m = j = 0;
        sum = 0;
        while (n < 100) {
            n++;
            m = 0;
            while (m < 100) {
                m++;
                j = 0;
                while (j < 10) {
                    j++;
                    sum += (n + m + j) / 2;
                }
            }
        }
        blackhole.consume(sum);
        blackhole.consume(n);
        blackhole.consume(m);
        blackhole.consume(j);
    }

    @Benchmark
    public void BenchmarkWhileD4(Blackhole blackhole) {
        int n, m, j, k;
        n = m = j = k = 0;
        sum = 0;
        while (n < 100) {
            n++;
            m = 0;
            while (m < 10) {
                m++;
                j = 0;
                while (j < 10) {
                    j++;
                    k = 0;
                    while (k < 10) {
                        k++;
                        sum += (n + m + j + k) / 2;
                    }
                }
            }
        }
        blackhole.consume(sum);
        blackhole.consume(n);
        blackhole.consume(m);
        blackhole.consume(j);
        blackhole.consume(k);
    }

    public static void main(String[] args) throws RunnerException {
        Options opt = new OptionsBuilder()
                .include(BenchmarkWhile.class.getSimpleName())
                .forks(1)
                .build();
        new Runner(opt).run();
    }
}
